server/app/cloud.go

```diff
@@ -29,3 +33,52 @@ func (a *App) GetBoardsCloudLimits() (*model.BoardsCloudLimits, error) {
 
 	return boardsCloudLimits, nil
 }
+
+func (a *App) NotifyPortalAdminsUpgradeRequest(workspaceID string) error {
+	if a.pluginAPI == nil {
+		return ErrNilPluginAPI
+	}
+
+	team, err := a.store.GetWorkspaceTeam(workspaceID)
+	if err != nil {
+		return err
+	}
+
+	var ofWhat string
+	if team == nil {
+		ofWhat = "your organisation"
+	} else {
+		ofWhat = team.DisplayName
+	}
+
+	message := fmt.Sprintf("A member of %s has notified you to upgrade this workspace before the trial ends.", ofWhat)
+
+	page := 0
+	getUsersOptions := &mmModel.UserGetOptions{
+		Active:  true,
+		Role:    mmModel.SystemAdminRoleId,
+		PerPage: 50,
+		Page:    page,
+	}
+
+	receiptUserIDs := []string{}
+
+	for ; true; page++ {
+		getUsersOptions.Page = page
+		systemAdmins, appErr := a.pluginAPI.GetUsers(getUsersOptions)
+		if appErr != nil {
+			a.logger.Error("failed to fetch system admins", mlog.Int("page_size", getUsersOptions.PerPage), mlog.Int("page", page), mlog.Err(appErr))
+			return appErr
+		}
+
+		if len(systemAdmins) == 0 {
+			break
+		}
+
+		for _, systemAdmin := range systemAdmins {
+			receiptUserIDs = append(receiptUserIDs, systemAdmin.Id)
+		}
+	}
+
+	return a.store.SendMessage(message, "custom_cloud_upgrade_nudge", receiptUserIDs)
```

Comment:
Shouldn't you send this per page. Otherwise, I'm not sure I understand such low per page number. Shouldn't be more than 50 admins, but what is paging protecting?

Reply:
I've now moved the sendMessage call to be per page.

Paging isn't needed here but the plugin API to get multiple users is paginated, so we either query a really large number in `PerPage` or check for multiple pages. I didn't wanna query a large number to avoid bringing so much data into in-memory, so did the page loop.